mypy/fastparse.py

```diff
@@ -520,13 +520,23 @@ def do_func_def(self, n: Union[ast3.FunctionDef, ast3.AsyncFunctionDef],
                 # Before 3.8, [typed_]ast the line number points to the first decorator.
                 # In 3.8, it points to the 'def' line, where we want it.
                 lineno += len(n.decorator_list)
+                end_lineno = None
+            else:
+                # It's okay that end_lineno is *before* lineno here.
+                # lineno is where the error is reported, and end_lineno defines
+                # the opposite end of the range to search for "# type: ignore"
+                # comments in. This reports the correct line in the event of an
+                # error, but keeps old pre-3.8 "# type: ignore" comments working
+                # correctly.
```

Comment:
This comment confused me. The key thing to note here is that we want to point `lineno` to the line that has the `def` keyword, while we want to point `end_lineno` to the line that would have been used for Python 3.7 and before. And we must compute `end_lineno` (i.e. the "old start line") using the crufty way given here since the old way of computing it (i.e. `n.lineno + len(n.decorator_list)` won't have the same result in 3.8.

Sorry that this is still so confusing. The code is right!

Reply:
Are you going to update these comments? I hope so!

Reply:
I'll try to reword this to be more concise. Maybe something like:

```py
# Set end_lineno to the old pre-3.8 lineno, in order to keep
# existing "# type: ignore" comments working:
```

Reply:
Sounds good -- also for the other location!